Stms to use gates

bashtanov opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
c/tx_gateway_frontend: don not grab STM read lock twice

bashtanov committed:
c/tx_gateway_frontend: hold tm_stm gate every time we use it

use `with_stm` helper for it

bashtanov committed:
c/tx_gateway_frontend: hold frontend gate in `delete_partition_from_tx`

While currently this function is only used by Admin API, which is
stopped before `tx_gateway_frontend`, for consistency we should make it
hold a gate so that `tx_gateway_frontend::stop()` shuts down cleanly.

bashtanov committed:
c/tm_stm: tidy-up

1) remove trivial override for start()
2) make `update_tx()` private

bashtanov committed:
c/tm_stm: do not hold gate

STM gate is held by tx_gateway_frontend, as it also holds locks that
belong to the STM

bashtanov commented:
/dt

bashtanov commented:
/dt

bashtanov committed:
c/tests/{rm_stm,idempotency}: reduce code duplication

also benefit from the fixture being friend to the stm class to reduce
boilerplate in the STM itself in the next commit

bashtanov commented:
/dt

bashtanov committed:
c/partition_manager: formatting

bashtanov commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d62d43ee-3493-41d7-8593-6bb75d731563-->
<!--TEST RESULTS-->
<!--4050d692-11aa-4343-863c-5db5ed000ae5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77115">build#77115</a></summary>


| test_class                          | test_method         | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                     |
|:------------------------------------|:--------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest             | test_simple_remount | {"cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/77115#019acd05-7773-4d08-90e9-b7d64ed7c5a5 | FLAKY         | 14/21    | upstream reliability is '80.11494252873564'. current run reliability is '66.66666666666666'. drift is 13.44828 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount        |
| WriteCachingFailureInjectionE2ETest | test_crash_all      | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/77115#019acd05-776f-44fb-bb8f-ce01fc09de82 | FLAKY         | 20/21    | upstream reliability is '90.56603773584906'. current run reliability is '95.23809523809523'. drift is -4.67206 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77138">build#77138</a></summary>


| test_class                   | test_method         | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                  |
|:-----------------------------|:--------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest      | test_simple_remount | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/77138#019ada39-ad03-42eb-a075-f43ab0d83b0e | FLAKY         | 17/21    | upstream reliability is '80.33775633293125'. current run reliability is '80.95238095238095'. drift is -0.61462 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount     |
| PartitionMovementUpgradeTest | test_basic_upgrade  | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/77138#019ada39-ad08-46ea-bc10-3c33c80e2368 | FLAKY         | 20/21    | upstream reliability is '99.71264367816092'. current run reliability is '95.23809523809523'. drift is 4.47455 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementUpgradeTest&test_method=test_basic_upgrade |
</details>
<!--/body-->

bashtanov committed:
raft/persisted_stm and implementations: hold the gate where necessary

In async member functions of STM implementations hold the gate to make
sure the STM survives the call.

Some of these changes are just defensive, as all existing callers of
affected functions are coroutines holding shared pointers to the STM.
Others fix genuine UBs.

Some functions were not added a gate holder or even had one removed:
- functions called internally by the STM machinery (see the comment at
persisted_stm::_gate for details);
- where lifetime is maintained by other means (e.g. those returning a
lock holder so that the caller needs to guarantee even longer lifetime,
see individual comments);
- in private functions, as lifetime is maintained by their callers.

bashtanov committed:
raft/persisted_stm implementations: remove trivial start/stop overrides

joe-redpanda commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -2694,48 +2694,40 @@ tx_gateway_frontend::get_all_transactions_for_one_tx_partition(
+                  return stm->read_lock().then([stm](
+                                                 ss::basic_rwlock<>::holder
+                                                   unit) {
+                      return stm->get_all_transactions()
```
q: is it assumed that 'get_all_transactions' should always happen behind the read lock?

bashtanov replied:
I assume so, as it iterates over `_transactions` which can be changed under write lock.

joe-redpanda replied:
wondering if its worth enforcing by having the function signature take the holders in and pass them back on await

bashtanov replied:
`get_all_transactions` is only called from here, so the easiest way to enforce locking would be to grab the lock inside. But this kind of rework is out of scope of this change, I'm only tidying up gates usage.

joe-redpanda commented on src/v/cluster/tests/idempotency_tests.cc:
```diff
@@ -20,14 +20,7 @@
 FIXTURE_TEST(
   test_rm_stm_doesnt_interfere_with_out_of_session_messages,
   rm_stm_test_fixture) {
-    create_stm_and_start_raft();
```
my favorite kind of commit

bharathv commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -2694,48 +2694,40 @@ tx_gateway_frontend::get_all_transactions_for_one_tx_partition(
-                    })
-                  .finally([u = std::move(unit)] {});
-            })
+          return self
```
gate_lock is a confusing name :)

bharathv commented on src/v/cluster/tm_stm.cc:
```diff
@@ -106,8 +106,7 @@ tm_stm::get_tx(kafka::transactional_id tx_id) {
 
 ss::future<checked<model::term_id, tm_stm::op_status>> tm_stm::barrier() {
-    return handle_shutdown_exceptions(
-      ss::with_gate(_gate, [this] { return do_barrier(); }));
```
I feel holding the stm gate inside the stm is probably the right way to do. Currently (unrelated to your change) its a bit odd that tx_gateway_frontend holds the stm gate. its a bigger refactor so :shrug:

bashtanov replied:
yeah these components are entangled more than appropriate

bharathv reviewed (approved):
lgtm
